### DISCUSSION TITLE
Add `lazy` function

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#264] [Ideas]
Define a property whose value isn't calculated until first access.

```ts
lazy(obj, 'key', () => expensiveCalculation())

obj.key // calculates the value on-demand
obj.key // returns a cached result on future access

// use `this` to access other properties
lazy(obj, 'key', function() {
  return expensiveCalculation(this.otherKey)
})
```

It can also be used as a decorator.

```ts
class MyClass {
  @lazy get thing() {
    return expensiveCalculation(this)
  }
}
```

**Comment by Minhir** [comment]
It can be easily solved with `once`:

```js
class MyClass {
  get thing() {
    return once(expensiveCalculation);
  }
}
```

**Reply by aleclarson** [reply]
I think you mean this?

```ts
class MyClass {
  getThing = once(expensiveCalculation.bind(this));
}
```

But yeah, probably best to avoid having multiple ways to accomplish the same thing. There's a downside in terms of stylistic preference, but that's not really a good enough justification. Arguably, keeping “expensive calculations” as functions is better anyway, since basic property access is often assumed to be cheap. Also, using `once` gives you support for `once.reset`, allowing you to clear the memoization.

Overall, I concur that this proposal should be rejected!
